lib/src/generate/build.dart

```diff
@@ -45,12 +51,92 @@ Future<BuildResult> build(List<List<Phase>> phaseGroups,
   writer ??=
       new CachedAssetWriter(cache, new FileBasedAssetWriter(packageGraph));
 
+  var buildImpl = new BuildImpl(
+      cache, new AssetGraph(), reader, writer, packageGraph, phaseGroups);
+
   /// Run the build!
-  var result = await new BuildImpl(
-          new AssetGraph(), reader, writer, packageGraph, phaseGroups)
-      .runBuild();
+  var futureResult = buildImpl.runBuild();
 
-  await logListener.cancel();
+  // Handle the first SIGINT, and shut down gracefully.
+  terminateEventStream ??= ProcessSignal.SIGINT.watch();
+  int numEventsSeen = 0;
+  var terminateListener = terminateEventStream.listen((_) {
+    numEventsSeen++;
+    if (numEventsSeen == 1) {
+      new Logger('Build').info('Waiting for build to finish...');
+    } else {
+      exit(2);
+    }
+  });
 
-  return result;
+  return futureResult.then((result) async {
+    await logListener.cancel();
```

Comment:
nit: do we need to wait for `logListener.cancel()` before canceling `terminateListener`? If not, consider:

```dart
await Future.wait([logListener.cancel(), terminateListener.cancel()]);
return result;
```